FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.JobManagement/AzureStorageQueueClient.cs

```diff
@@ -0,0 +1,682 @@
+            EnsureArg.IsNotNull(storage, nameof(storage));
+
+            _logger = EnsureArg.IsNotNull(logger, nameof(logger));
+
+            if (storage.UseConnectionString)
+            {
+                _azureJobInfoTableClient = new TableClient(storage.TableUrl, storage.TableName);
+                _azureJobMessageQueueClient = new QueueClient(storage.QueueUrl, storage.QueueName);
+            }
+            else
+            {
+                _azureJobInfoTableClient = new TableClient(
+                    new Uri(storage.TableUrl),
+                    storage.TableName,
+                    new DefaultAzureCredential());
+                _azureJobMessageQueueClient = new QueueClient(
+                    new Uri($"{storage.QueueUrl}{storage.QueueName}"),
+                    new DefaultAzureCredential());
+            }
+        }
+
+        public bool IsInitialized()
+        {
+            try
+            {
+                _azureJobInfoTableClient.CreateIfNotExists();
+                _azureJobMessageQueueClient.CreateIfNotExists();
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "Failed to initialize azure storage client.");
+                return false;
+            }
+
+            _logger.LogInformation("Initialize azure storage client successfully.");
+            return true;
+        }
+
+        // The expected behaviors:
+        // 1. multi agent instances enqueue same jobs concurrently, only one job entity will be created,
+        //    there may be multi messages, while only one will be recorded in job lock entity, and the created jobInfo will be returned for all instances.
+        // 2. re-enqueue job, no matter what the job status is now, will do nothing, and return the existing jobInfo.
+        // 3. if one of the steps fails, will continue to process it when re-enqueue.
+        // TODO: The parameter forceOneActiveJobGroup and isCompleted are ignored for now
+        public async Task<IEnumerable<JobInfo>> EnqueueAsync(
+            byte queueType,
+            string[] definitions,
+            long? groupId,
+            bool forceOneActiveJobGroup,
+            bool isCompleted,
+            CancellationToken cancellationToken)
+        {
+            _logger.LogInformation($"Start to enqueue {definitions.Length} jobs.");
+
+            // step 1: get incremental job ids, will try again if fails
+            var jobIds = await GetIncrementalJobIds(queueType, definitions.Length, cancellationToken);
+
+            // handle for each job
+            var jobInfos = new List<TJobInfo>();
+            for (var i = 0; i < definitions.Length; i++)
+            {
+                Trace.Assert(
+                    definitions[i].Length * sizeof(char) < MaximumSizeOfAnEntityInBytes,
+                    "The maximum size of a single table entity is 1MB, the size of definition is larger than 1MB.");
```

Comment:
We'd better add log here.

Reply:
Log an error and throw exceptions for this case.